lib/cache/memory-cache-store.js

```diff
@@ -0,0 +1,133 @@
+'use strict'
+
+/**
+ * @typedef {import('../../types/cache-interceptor.d.ts').default.CacheStore} CacheStore
+ * @implements {CacheStore}
+ */
+class MemoryCacheStore {
+  /**
+   * @type {import('../../types/cache-interceptor.d.ts').default.MemoryCacheStoreOpts} opts
+   */
+  #opts = {}
+  #entryCount = 0
+  /**
+   * @type {Map<string, import('../../types/cache-interceptor.d.ts').default.CacheStoreValue[]>}
+   */
+  #data = new Map()
+
+  /**
+   * @param {import('../../types/cache-interceptor.d.ts').default.MemoryCacheStoreOpts | undefined} opts
+   */
+  constructor (opts) {
+    this.#opts = opts ?? {}
+
+    if (!this.#opts.maxEntries) {
+      this.#opts.maxEntries = Infinity
+    }
+
+    if (!this.#opts.maxEntrySize) {
+      this.#opts.maxEntrySize = Infinity
+    }
+  }
+
+  get entryCount () {
+    return this.#entryCount
+  }
+
+  get maxEntries () {
+    return this.#opts.maxEntries
+  }
+
+  get maxEntrySize () {
+    return this.#opts.maxEntrySize
+  }
+
+  /**
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @returns {import('../../types/cache-interceptor.d.ts').default.CacheStoreValue | undefined}
+   */
+  get (req) {
+    const key = this.#makeKey(req)
+    if (!key) {
+      return undefined
+    }
+
+    const values = this.#data.get(key)
+    if (!values) {
+      return
+    }
+
+    let value
+    const now = Date.now()
+    for (let i = values.length - 1; i >= 0; i--) {
+      const current = values[i]
+      if (now >= current.deleteAt) {
+        // Should be deleted, so let's remove it
+        values.splice(i, 1)
+        this.#entryCount--
+        continue
+      }
+
+      let matches = true
+
+      if (current.vary) {
+        for (const key in current.vary) {
+          if (current.vary[key] !== req.headers[key]) {
```

Comment:
This looks like a horrible algorithm. First of all, you could iterate over all entries and from 0 to length and if you find an entry you want to delete you move the last entry to that position and reduce the entryCount. At the end you just set the length of the array to the entryCount and by this delete now the duplicated entries.

And secondly you could insert the entries sorted to the right place. Like sorting them from big to small, then you would only need to iterate from the right to left, and then you just cut of after the first element which is not to be deleted.

E.g. it is now 88ms, and you want to cut of the older ones. If it is sorted like this:
```js
const values = [
  {deleteAt: 100},
  {deleteAt: 100},
  {deleteAt: 90},
  {deleteAt: 80},
  {deleteAt: 80},
  {deleteAt: 70},
]
```

Then you only need to do `values.length = 3` to remove the three entries